designs/0028-parallel_map.md

```diff
@@ -0,0 +1,129 @@
+- Feature Name: parallel_map
+- Start Date: 2020/09/29
+- RFC PR: (leave this empty)
+- Stan Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+The proposed ```parallel_map``` functionality provides a means of within-chain parallelisation that does not require the sum of iterations be returned (i.e., ```reduce_sum```) or any complex argument-packing (i.e., ```map_rect```).
+
+# Motivation
+[motivation]: #motivation
+
+While the ```reduce_sum``` framework has been very popular for its ease of use, it's limited to situations where the goal is to return the sum of independent computations. Many users have models with large functions/computations that return vectors or matrices in full, rather than their sum. For these cases, users seeking parallelisation are limited to the ```map_rect``` functionality, which has a non-zero barrier to entry due to the need for argument packing and unpacking. The ```parallel_map``` framework aims to provide a means of parallelising these operations in a manner which is easy for users to understand and implement.
```

Comment:
> large functions/computations that return vectors or matrices in full

Returning anything but scalars and I'm gonna complain :P. Precomputing gradients with lots of outputs is not efficient with reverse mode. To do that efficiently will require parallel reverse mode.

Of course, if you want to do mess with parallel reverse mode I'm all for it!

Reply:
> Precomputing gradients with lots of outputs is not efficient with reverse mode

Benchmarking has shown that it's definitely not great in all cases, but it does provide some speed-up. I still need to add some more robust benchmarking to this proposal though.

> Of course, if you want to do mess with parallel reverse mode I'm all for it!

Haha would not know where to start! I'd be interested in tinkering with it though. Any chance there's a writeup somewhere of where the parallelism needs to go in? Or if you're feeling particularly generous could we set up a hangouts to go over the current autodiff and what needs to be built out for parallel reverse mode?

Reply:
Just found @wds15 [design doc](https://github.com/wds15/design-docs/blob/parallel_autodiff/designs/0003-parallel_autodiff.md) and [math branch](https://github.com/stan-dev/math/compare/parallel-ad-tape-3) for the parallel autodiff, will start reading up on those

Reply:
> Benchmarking has shown that it's definitely not great in all cases

The problem with multiple outputs (and the reason I'm so allergic to it), is I'm scared it makes it possible for people to write slow code.

What we're doing is computing the Jacobian of a function `f` using reverse mode autodiff and then embedding that in a forward pass.

On paper, if you have a function with N inputs and M outputs, the Jacobian is an MxN matrix and it takes M reverse mode passes or N forward mode passes to compute it.

In this case, parallel_map is a bunch of evaluations of a little function `g`, so something like $f = [g_1, g_2, ...]$

Because each of the `g` functions in the parallel_map calculation don't share calculations, we can separate out the adjoints on the input, and we can do all these M reverse mode passes in parallel.

If each of the `g` functions has multiple outputs though, those will not be parallelizable and you'll need to do multiple reverse mode passes.

The caveat to this is if there is a function where the forward pass is slower than the reverse pass, you can get away with multiple reverse passes because they are cheap. The ODE solvers are currently like this, all the distributions are like this, and some of the linear algebra stuff has different complexity for the forward/reverse passes.

I don't like it because someone needs to guess and check whether their model will be faster and they can accidentally write code that is slower than the serial version (not due to parallel overhead -- there's just more work in computing all the Jacobians). This is how we ended up with the current single-output focused `reduce_sum` design.

> would not know where to start

The problem with parallel reverse mode is how to handle race conditions if multiple `chain` functions are being evaluated in parallel, so that's where to start thinking. Yeah we can have a meeting whenever. @wds15 probably would like in too.

The larger problem of parallel is a lot of models are memory bound, and so we won't get faster unless we do something parallel across multiple computers. It's not something that needs written at first, but I want to keep multiple-computers in mind for the design.

Reply:
Thanks for writing that all out, super helpful. I'll mull this over for couple days and try to get my head as far around it as I can. Maybe hang around after the general meeting? @wds15 work for you?

> The larger problem of parallel is a lot of models are memory bound, and so we won't get faster unless we do something parallel across multiple computers. It's not something that needs written at first, but I want to keep multiple-computers in mind for the design.

I don't know if we'll get that through the TBB, so that might need a different implementation all together

Reply:
> Maybe hang around after the general meeting?

Works for me

> through the TBB

Yeah it wouldn't be the TBB.

Reply:
My apologies for not yet having digged into this, but it sounds like the main points in my mind are already hammered on (in a good way):

- The index function concept is cool and if we can make it happen here we should think about putting it into `reduce_sum_v2` or whatever.
- The index function thing should ideally also work for ranges.
- What is unclear to me is how the index function stuff can actually work with the Stan language... and in case the index function thing is a major project on its own, then maybe the prospect of tuples makes the need for an index function obsolete (with compromises)?

I do think that the return type should be anything - even though users can easily slow down their program due to the large Jacobians. There are sufficient applications where the memory boundness is not a problem - like ODE solving, mixture models, joint GP's where multiple covariance matrices are constructed (if that is done in practice).

For truly parallel reverse mode AD, one would probably come up with something which leverages information from the compiler in order to know upfront how to manage the synchronization.

On Thursday after the Stan meeting should work.